reconcile/ocm_internal_notifications/integration.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from collections.abc import (
+    Callable,
+    Iterable,
+)
+from functools import lru_cache
+
+from reconcile.gql_definitions.common.ocm_environments import (
+    query as ocm_environment_query,
+)
+from reconcile.gql_definitions.fragments.ocm_environment import OCMEnvironment
+from reconcile.slack_base import slackapi_from_queries
+from reconcile.utils import gql
+from reconcile.utils.ocm_base_client import (
+    OCMAPIClientConfigurationProtocol,
+    OCMBaseClient,
+    init_ocm_base_client,
+)
+from reconcile.utils.runtime.integration import (
+    NoParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import SecretReaderBase
+
+QONTRACT_INTEGRATION = "ocm-internal-notifications"
+
+
+class OcmInternalNotifications(QontractReconcileIntegration[NoParams]):
+    """Something."""
+
+    def __init__(self) -> None:
+        super().__init__(NoParams())
+        self.slack = slackapi_from_queries(
+            integration_name=self.name, init_usergroups=False
+        )
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_environments(self, query_func: Callable) -> list[OCMEnvironment]:
+        return ocm_environment_query(query_func).environments
+
+    def init_ocm_apis(
+        self,
+        environments: Iterable[OCMEnvironment],
+        init_ocm_base_client: Callable[
+            [OCMAPIClientConfigurationProtocol, SecretReaderBase], OCMBaseClient
+        ] = init_ocm_base_client,
+    ) -> dict[str, OCMBaseClient]:
+        return {
+            env.name: init_ocm_base_client(env, self.secret_reader)
+            for env in environments
+        }
+
+    @lru_cache
+    def slack_get_user_id_by_name(self, user_name: str, mail_address: str) -> str:
+        return self.slack.get_user_id_by_name(
+            user_name=user_name, mail_address=mail_address
+        )
+
+    def run(self, dry_run: bool) -> None:
+        gqlapi = gql.get_api()
+        environments = self.get_environments(gqlapi.query)
+
+        self.ocm_apis = self.init_ocm_apis(environments, init_ocm_base_client)
+
+        for env_name, ocm in self.ocm_apis.items():
+            if env_name == "ocm-production":
+                continue
+
+            clusters = ocm.get(
+                api_path="/api/clusters_mgmt/v1/clusters",
+                params={"search": "state like 'uninstalling' and managed='true'"},
+            ).get("items")
+
+            slack_user_ids = set()
+            for cluster in clusters:
+                subscription = ocm.get(api_path=cluster["subscription"]["href"])
+                creator = ocm.get(api_path=subscription["creator"]["href"])
+                email = creator["email"]
+                user, mail_address = email.split("@")
+                user_name = user.split("+")[0]
```

Comment:
`user` might already contain what you are looking for. not everyone uses gmail aliases for their OCM account. this might be a quick win to increase the hit rate for finding an adressable slack user.

some optimizations for a later stage
* RH associates can request arbitrary email aliases. maybe they are stored in LDAP and can help to find the slack handle
* cluster creators can be service accounts as well. in such cases we could look into the organization and find their members/administrators and find a something that resolves to a slack handle there
* /api/accounts_mgmt/v1/subscriptions/{subId}/notification_contacts
* for well known big organizations with high throughput we could place some contacts in org labels (demo.redhat.com)

Reply:
since this is internal environments only, we are expecting mostly internal users not to RH, but to SD.

the current code covers all the existing clusters in the searched state, and i'm happy to fail forward.

Reply:
ah understood. ack